modules/transactions/src/main/java/org/apache/ignite/internal/tx/impl/TxManagerImpl.java

```diff
@@ -439,20 +472,26 @@ private CompletableFuture<Void> durableFinish(
                     if (ex != null) {
                         Throwable cause = ExceptionUtils.unwrapCause(ex);
 
-                        if (cause instanceof TransactionException) {
-                            TransactionException transactionException = (TransactionException) cause;
+                        if (cause instanceof TransactionAlreadyFinishedException) {
+                            TransactionAlreadyFinishedException transactionException = (TransactionAlreadyFinishedException) cause;
+
+                            TransactionResult result = transactionException.transactionResult();
 
-                            if (transactionException.code() == TX_WAS_ABORTED_ERR) {
-                                updateTxMeta(txId, old -> {
-                                    TxStateMeta txStateMeta = new TxStateMeta(ABORTED, old.txCoordinatorId(), commitPartition, null);
+                            updateTxMeta(txId, old -> {
+                                TxStateMeta txStateMeta =
+                                        new TxStateMeta(
+                                                result.transactionState(),
+                                                old.txCoordinatorId(),
+                                                commitPartition,
+                                                result.commitTimestamp()
+                                        );
 
-                                    txFinishFuture.complete(txStateMeta);
+                                txFinishFuture.complete(txStateMeta);
```

Comment:
By the way, let's better move this future completion out of the meta updater (and hence from the map.compute closure)